src/OrbitDB.js

```diff
@@ -266,34 +267,43 @@ class OrbitDB {
   }
 
   // Callback when database was closed
-  async _onClose (address) {
+  async _onClose (db) {
+    const address = db.address.toString()
+
     logger.debug(`Close ${address}`)
 
     // Unsubscribe from pubsub
     if (this._pubsub) {
       await this._pubsub.unsubscribe(address)
     }
 
-    // Cache cleanup
-    // First remove the handler by db address
-    // Then close the cache if there are no more handlers
-    for(let directory of Object.keys(this.caches)) {
-      const cache = this.caches[directory] || this.cache
-      const store = this.stores[address]
-
-      if (store && store.options.directory == directory) {
-        // Closing a custom cache?
-        if (cache.handlers.has(address)) cache.handlers.delete(address)
-        if (!cache.handlers.size) await cache.close()
-      } else {
-        // Else remove it from the default cache
-        if (this.cache.handlers.has(address)) this.cache.handlers.delete(address)
-        if (!this.cache.handlers.size) await this.cache.close()
+    const store = this.stores[address]
+
+    if (store && store.options.directory) {
+      const cache = this.caches[store.options.directory]
+      // Closing a custom cache?
+      if (cache && cache.handlers.has(address.toString())) {
+        cache.handlers.delete(address.toString())
+        if (!cache.handlers.size) await cache.cache.close()
+      }
+
+      if (this.caches[this.directory].handlers.has(address.toString())) {
+        this.caches[this.directory].handlers.delete(address.toString())
+        if (!this.caches[this.directory].handlers.size) await this.cache.close()
       }
     }
+
     delete this.stores[address]
   }
 
+  async _onDrop (db) {
+    await db._cache.open()
```

Comment:
Why open on drop again?

Reply:
Because there are cache.del calls inside of Store.drop that need the cache to be open. If Store.close is called before Store.drop like in the persistency tests, this will happen.